Create ServerNotification Components

labkey-danield opened the pull request:
#### Rationale
This is the test component for the ServerNotification component that is in the Biologics and SampleManager apps. This is not the same as notifications in LKS.

#### Related Pull Requests
* https://github.com/LabKey/sampleManagement/pull/738/files

#### Changes
* Created menu component (ServerNotificationMenu.java)
* Created notification item component (ServerNotificationItem.java)
* Added menu to the NavBar.

### Overview
The server notification menu item is the little "bell" icon in the nav bar in Biologics and SampleManager.
<img width="487" alt="Screen Shot 2021-10-18 at 5 27 39 PM" src="https://user-images.githubusercontent.com/12738200/137835629-b5a59bc3-7865-45ce-8e3f-ce2a62f7f9fd.png">

It gives a notification / status of the background imports. 

Before any imports happen the list is empty:
<img width="481" alt="Screen Shot 2021-10-18 at 5 36 09 PM" src="https://user-images.githubusercontent.com/12738200/137835777-d23c1496-8fb3-4b91-9462-f37f29a940fa.png">

When an import job is running the icon changes from a bell to a spinner:
<img width="533" alt="Screen Shot 2021-10-18 at 5 36 34 PM" src="https://user-images.githubusercontent.com/12738200/137835896-382bac6a-bec7-4b57-9da2-e4c8ce94973b.png">

If there are unread messages a "badge" is shown next to the bell icon indicating the number of unread messages. Click the bell will show the list of messages (read and unread).
<img width="509" alt="Screen Shot 2021-10-18 at 5 28 13 PM" src="https://user-images.githubusercontent.com/12738200/137836020-9328731c-d4f7-41a5-987a-78ba039d3b9a.png">

The functionality just described is covered by the ServerNotificationMenu class.
_Note: I forgot to add a method to cover the "Mark all as read" action._

As more imports are done, more items are added to the list. When all notifications are read the badge is removed from the bell:
<img width="522" alt="Screen Shot 2021-10-18 at 7 58 13 PM" src="https://user-images.githubusercontent.com/12738200/137836424-a4baa970-3058-4aa7-84da-519d56670215.png">

Each of the items in the list is covered by the ServerNotificationItem class.
Each entry in the list will have a status icon, a message, a 'View' link and a timestamp.
The 'View' link can go to different location depending upon the message (success or fail) and the app. Because of this the click method for the view link does not assume where the navigation will go.

If the import has an error the icon will be different:
<img width="488" alt="Screen Shot 2021-10-18 at 8 01 20 PM" src="https://user-images.githubusercontent.com/12738200/137837027-394c0348-68da-41dd-974e-265f0507642a.png">

Finally, it is possible to expand this menu while a job is running. In that case some of the components of the list item will not be present:
<img width="485" alt="Screen Shot 2021-10-18 at 5 43 55 PM" src="https://user-images.githubusercontent.com/12738200/137837154-c9e4fdf2-453f-4c1d-8040-0fc79f30269b.png">

As you can see from the image the 'View' link is not yet present but a username is (the username goes away when the job is completed). Also as you can see the icon changes to a spinner.

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Add AppNotification component.
Add component to NavBar.

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Add support for the 'Mark all as read' link.

labkey-tchad commented on src/org/labkey/test/components/ui/navigation/apps/ServerNotificationItem.java:
```diff
@@ -0,0 +1,218 @@
+    }
+
+
+    public static class AppNotificationEntryFinder extends WebDriverComponentFinder<ServerNotificationItem, AppNotificationEntryFinder>
```
With such a simple locator, I'd make this package protected so that only the `ServerNotificationMenu` can access it.
```suggestion
   static class AppNotificationEntryFinder extends WebDriverComponentFinder<ServerNotificationItem, AppNotificationEntryFinder>
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/navigation/apps/ServerNotificationItem.java:
```diff
@@ -0,0 +1,218 @@
+package org.labkey.test.components.ui.navigation.apps;
```
This should match the component package.
```suggestion
package org.labkey.test.components.ui.notifications;
```

labkey-danield replied:
Done.

labkey-tchad commented on src/org/labkey/test/components/ui/navigation/apps/ServerNotificationMenu.java:
```diff
@@ -0,0 +1,201 @@
+package org.labkey.test.components.ui.navigation.apps;
```
```suggestion
package org.labkey.test.components.ui.notifications;
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/navigation/apps/ServerNotificationMenu.java:
```diff
@@ -0,0 +1,201 @@
+        public final WebElement statusIcon()
+        {
+            return Locator.tagWithClass("i", "fa").findWhenNeeded(this);
+        }
```
This locator could be a bit more specific.
```suggestion
        public final WebElement statusIcon()
        {
            return Locator.byClass("navbar-header-icon").findWhenNeeded(this);
        }
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/navigation/apps/ServerNotificationMenu.java:
```diff
@@ -0,0 +1,201 @@
+        public final WebElement markAll()
+        {
+            return Locator.tagWithClass("h3", "navbar-menu-header")
+                    .child("div[contains(@class,'server-notifications-link')]").findWhenNeeded(this);
```
Using locator builders will generate CSS Locators which are, generally, faster that XPath.
```suggestion
            return Locator.tagWithClass("h3", "navbar-menu-header")
                    .child(Locator.tagWithClass("div", "server-notifications-link").findWhenNeeded(this);
```

labkey-danield replied:
Changed.
I tried to avoid the XPath but couldn't get the right locator. Thanks!

labkey-danield committed:
Changed package component is in.
Changed scope of the finder for ServerNotificationItem.
Misc other code review feedback.

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Add assay notification tests.
Added clickImportAsync method to handle an async run import.
Fix typo of success message.
Add a PipelineStatusPage to verify error imports.

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Consolidated tests into one test class.
Added test for sources.

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Added a few more UI checks to the tests.

labkey-danield committed:
Modified ServerNotificationMenu to return only one message identified by position in list or text of message.
Updated finder for ServerNotificationItem to be able to search based on message text or position.
Updated comments to be consistent when referring to a notification item vs. the message part of the item.

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-tchad commented on src/org/labkey/test/components/ui/notifications/ServerNotificationMenu.java:
```diff
@@ -0,0 +1,266 @@
+     */
+    public void clickMarkAll()
+    {
+        if(!isExpanded())
+        {
+            expand();
+        }
```
This `if` is redundant. `expand()` checks whether the menu is expanded.
```suggestion
        expand();
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/notifications/ServerNotificationMenu.java:
```diff
@@ -0,0 +1,266 @@
+    private WebElement waitForNotificationList()
+    {
+
+        if(!isExpanded())
+            expand();
```
```suggestion
        expand();
```

labkey-danield replied:
Changed.

labkey-tchad commented on src/org/labkey/test/components/ui/notifications/ServerNotificationMenu.java:
```diff
@@ -0,0 +1,266 @@
+    public String getNoNotificationsMessage()
+    {
+
+        if(!isExpanded())
+            expand();
```
```suggestion
        expand();
```

labkey-danield replied:
Changed.

labkey-danield committed:
Merge branch 'develop' into fb_smAsyncImportAutomation

labkey-danield committed:
Changed the test name from testRunningStatus to testActiveImportIndicator.
Added checks to that test to validate the spinner in the mega-menu and on a sample's overview page.
Implemented other code review feedback.
